vcGL/src/vcTextureHelper.cpp

```diff
@@ -343,18 +357,21 @@ udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, c
   UD_ERROR_NULL(pFrameBufferPixels, udR_MemoryAllocationFailure);
 
   vcTexture_BeginReadPixels(pTexture, 0, 0, currSize.x, currSize.y, pFrameBufferPixels, pFramebuffer);
-
   UD_ERROR_CHECK(vcTexture_ConvertPixels(pFrameBufferPixels, format, &pPixelsOut, vcTextureFormat_RGBA8, currSize.x * currSize.y));
+  
+  if (udStrEndsWithi(pFilename, "png"))
+    UD_ERROR_IF(stbi_write_png_to_func(vcTexture_StbCallback, &callbackResult, (int)currSize.x, (int)currSize.y, 4, pPixelsOut, 0) == 0, udR_InternalError);
+  else if (udStrEndsWithi(pFilename, "jpg"))
+    UD_ERROR_IF(stbi_write_jpg_to_func(vcTexture_StbCallback, &callbackResult, (int)currSize.x, (int)currSize.y, 4, pPixelsOut, 0) == 0, udR_InternalError);
+  else
+    UD_ERROR_SET(udR_Unsupported);
 
-  pWriteData = stbi_write_png_to_mem(pPixelsOut, 0, currSize.x, currSize.y, 4, &outLen);
-  UD_ERROR_NULL(pWriteData, udR_InternalError);
+  UD_ERROR_CHECK(callbackResult.result);
 
-  UD_ERROR_CHECK(udFile_Save(pFilename, pWriteData, outLen));
-  
   result = udR_Success;
 epilogue:
-  if (pWriteData != nullptr)
-    STBIW_FREE(pWriteData);
+
+  UD_ERROR_CHECK(udFile_Close(&callbackResult.pFile));
```

Comment:
This has the ability to create an infinite loop, and it has the ability to make this function appear as if it succeeded when it didn't. In general, don't use these macros or set `result` after the `epilogue`.